// @alexa-games/sfb-cli/dist/lib/filePath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utilities_1 = require("./utilities");
const pathModule = require('path');
class FilePath {
    constructor(path) {
        this.path = pathModule.resolve(path);
        this.delimiter = utilities_1.Utilities.isWin32 ? '\\' : '/';
        this.pathRoot = this.delimiter;
        const result = this.path.match(FilePath.rootRegEx);
        if (result && result.length > 0) {
            this.pathRoot = result[0];
        }
        if (!this.path.startsWith(this.pathRoot)) {
            throw new Error('Expected resolved path to start with ' + this.pathRoot);
        }
        const pathToSplit = path.substr(this.pathRoot.length);
        this.pathParts = pathToSplit.split(this.delimiter);
    }
    get length() {
        return this.pathParts.length;
    }
    getPartialPath(length) {
        if (length > this.pathParts.length) {
            throw new Error('Parameter length must not exceed length.');
        }
        if (length < 0) {
            throw new Error('Parameter length be at least 0.');
        }
        let result = this.pathRoot;
        for (let i = 0; i < length; i++) {
            if (i > 0) {
                result += this.delimiter;
            }
            result += this.pathParts[i];
        }
        return result;
    }
}
exports.FilePath = FilePath;
FilePath.rootRegEx = /^([A-Z]:\\)/;
