// @alexa-games/sfb-cli/dist/lib/testCommand.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const specialPaths_1 = require("./specialPaths");
const sfb_skill_1 = require("@alexa-games/sfb-skill");
class TestCommand {
    constructor(storyPath, locale, logger) {
        this.storyPath = storyPath;
        this.locale = locale;
        this.logger = logger;
    }
    async run() {
        this.logger.status('Testing content ...');
        if (!fs.existsSync(this.storyPath)) {
            throw new Error(`${this.storyPath}: The specified story path does not exist.`);
        }
        const dirs = new specialPaths_1.SpecialPaths(this.storyPath);
        const config = await sfb_skill_1.ConfigAccessor.loadConfigFile(dirs.abcConfig, dirs.builtResourcesPath);
        await this.testStory(this.locale, dirs, config);
    }
    testStory(locale, dirs, config) {
        const bakedFilePath = path.resolve(dirs.bakedPath, locale, config.getValue("abc-baked-filename", undefined, locale));
        const extensionLoaderModule = require(dirs.extensionLoaderPath);
        const customExtensions = new extensionLoaderModule.ExtensionLoader({
            locale,
            configAccessor: config
        });
        const customStoryExtensions = customExtensions.getExtensions();
    }
}
exports.TestCommand = TestCommand;
