// @alexa-games/sfb-f/dist/extensions/ACEExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isInstructionExtension(extension) {
    return true; // Since "InstructionExtension" doesn't have required methods, any object is "technically" an InstructionExtension
}
exports.isInstructionExtension = isInstructionExtension;
function isDriverExtension(extension) {
    return extension.pre !== undefined;
}
exports.isDriverExtension = isDriverExtension;
function isImporterExtension(extension) {
    return extension.extendImportedContent !== undefined;
}
exports.isImporterExtension = isImporterExtension;
var GlobalDirectionsExtension_1 = require("./coreExtensions/GlobalDirectionsExtension");
exports.GlobalDirectionsExtension = GlobalDirectionsExtension_1.GlobalDirectionsExtension;
var LocalizationExtension_1 = require("./coreExtensions/LocalizationExtension");
exports.LocalizationExtension = LocalizationExtension_1.LocalizationExtension;
var SnippetExtension_1 = require("./coreExtensions/SnippetExtension");
exports.SnippetExtension = SnippetExtension_1.SnippetExtension;
var VoiceOverExtension_1 = require("./coreExtensions/VoiceOverExtension");
exports.VoiceOverExtension = VoiceOverExtension_1.VoiceOverExtension;
