// @alexa-games/sfb-skill/dist/aceExtension/alexaAPLExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const APLHelper_1 = require("./../handler/APLHelper");
class AlexaAPLExtension {
    constructor(locale, configAccessor) {
        try {
            this.aplHelper = new APLHelper_1.APLHelper(locale, configAccessor);
        }
        catch (err) {
            throw {
                errorMessage: err,
                errorName: 'APL Extension Error',
                lineNumber: 0,
                sourceID: "APL_TEMPLATE"
            };
        }
    }
    async post(param) {
        const handlerInput = param.userInputHelper.getHandlerInput();
        const driver = param.driver;
        let visualProperties = await driver.getVisuals();
        if (handlerInput && visualProperties && this.aplHelper.supportsDisplay(handlerInput)) {
            let visualDirectives = this.aplHelper.generateAPLDirectiveWithVisualOptions(visualProperties);
            for (let aplDirective of visualDirectives) {
                handlerInput.responseBuilder.addDirective(aplDirective);
            }
        }
    }
    async pre(param) {
    }
}
exports.AlexaAPLExtension = AlexaAPLExtension;
