// @alexa-games/sfb-f/dist/test/integrationTests/import.int-spec.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const importer_1 = require("../../importer/importer");
const assert_1 = require("assert");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const DEFAULT_IMPORT_PLUGIN_NAME = "default";
const TEST_STORY_TITLE = "Test Story";
const TEST_STORY_ID = "test-id";
const TEST_BUILT_IN_COUNT = 23;
describe("Scene Transition Baseline", function () {
    let thinImporter;
    before(async function () {
        console.log("Reset ACEImporter...");
        thinImporter = new importer_1.ACEImporter([], undefined, []);
    });
    it("Import - split utterances into two groups", async function () {
        const testFileName = "utterance-splitting.abc";
        const importResult = await thinImporter.importABCStory(DEFAULT_IMPORT_PLUGIN_NAME, "", TEST_STORY_TITLE, TEST_STORY_ID, true, {
            contents: [
                {
                    id: testFileName,
                    text: fs.readFileSync(getTestResourcePath(testFileName), "utf8")
                }
            ]
        });
        const model = importResult.alexaVoiceModel;
        assert_1.strict.equal(model.languageModel.intents.length - TEST_BUILT_IN_COUNT, 2, "Unexpected number of intents for given test.");
    });
    it("Import - build up intent for addional utterance", async function () {
        const testFileName = "intent-build-up.abc";
        const importResult = await thinImporter.importABCStory(DEFAULT_IMPORT_PLUGIN_NAME, "", TEST_STORY_TITLE, TEST_STORY_ID, true, {
            contents: [
                {
                    id: testFileName,
                    text: fs.readFileSync(getTestResourcePath(testFileName), "utf8")
                }
            ]
        });
        const model = importResult.alexaVoiceModel;
        assert_1.strict.equal(model.languageModel.intents.length - TEST_BUILT_IN_COUNT, 1, "Unexpected number of intents for given test.");
    });
    it("Import - normal intent splitting per choice", async function () {
        const testFileName = "normal-intent-split-per-choice.abc";
        const importResult = await thinImporter.importABCStory(DEFAULT_IMPORT_PLUGIN_NAME, "", TEST_STORY_TITLE, TEST_STORY_ID, true, {
            contents: [
                {
                    id: testFileName,
                    text: fs.readFileSync(getTestResourcePath(testFileName), "utf8")
                }
            ]
        });
        const model = importResult.alexaVoiceModel;
        assert_1.strict.equal(model.languageModel.intents.length - TEST_BUILT_IN_COUNT, 3, "Unexpected number of intents for given test.");
    });
});
function getTestResourcePath(testResourceName) {
    return path.resolve(__dirname, "..", // ./dist/test
    "..", // ./dist,
    "..", // ./
    "test", // ./test
    "integrationTests", // ./test/integrationTests
    "testResources", // ./test/integrationTests/testResources
    testResourceName);
}
