// @alexa-games/sfb-f/dist/importPlugins/storyBlock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const storyFeatureEnumerator_1 = require("./storyFeatureEnumerator");
class StoryBlock {
    constructor(lines, blockName, range) {
        this.lines = lines;
        this.blockName = blockName;
        this.range = range;
    }
    /**
     * Assembles the whole block into a single multi-line string.
     */
    getText() {
        let result = undefined;
        ;
        for (let i = this.range.start.row; i <= this.range.end.row; i++) {
            if (result) {
                result += '\n';
            }
            else {
                result = '';
            }
            result += this.range.cropToRange(this.lines[i], i);
        }
        return result || '';
    }
}
exports.StoryBlock = StoryBlock;
class SceneStoryBlock extends StoryBlock {
    constructor(lines, blockName, range) {
        super(lines, blockName, range);
    }
    getPropertyEnumerator(type) {
        return new storyFeatureEnumerator_1.ScenePropertyEnumerator(this.lines, this.range, type);
    }
}
exports.SceneStoryBlock = SceneStoryBlock;
class ScenePropertyStoryBlock extends StoryBlock {
    constructor(lines, blockName, range) {
        super(lines, blockName, range);
    }
}
exports.ScenePropertyStoryBlock = ScenePropertyStoryBlock;
