// @alexa-games/sfb-f/dist/importPlugins/userInputHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Accessor/Helper for UserInput used as an input for the Driver's story processing.
 */
class UserInputHelper {
    constructor(userInput) {
        this.userInput = userInput;
    }
    /**
     * Gets the HandlerInput generated by the ASK SDK for the current request
     */
    getHandlerInput() {
        return this.userInput.handlerInput;
    }
    getInputValue() {
        return this.userInput.value;
    }
    setInputValue(value) {
        this.userInput.value = value;
    }
    getInputSlots() {
        return this.userInput.slots;
    }
    setInputSlots(slots) {
        this.userInput.slots = slots;
    }
    getInputIntent() {
        return this.userInput.intent;
    }
    setInputIntent(intent) {
        this.userInput.intent = intent;
    }
    addInputSlot(slotName, value) {
        if (!this.userInput.slots) {
            this.userInput.slots = [];
        }
        this.userInput.slots.push({
            name: slotName,
            value: value
        });
    }
    clearUserInput() {
        delete this.userInput.intent;
        delete this.userInput.slots;
        delete this.userInput.value;
    }
    getUserInput() {
        return this.userInput;
    }
}
exports.UserInputHelper = UserInputHelper;
