// @alexa-games/sfb-skill/dist/handler/SFBRequestHandlerFactory.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const AlexaSFBRequestHandler_1 = require("./AlexaSFBRequestHandler");
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
class SFBRequestHandlerFactory {
    static create(event, context, extensions, configAccessor, projectDir, customRequestHandler) {
        configAccessor.deployed = true;
        const locale = event.request.locale ? event.request.locale : "en-US";
        const dynamoDBTalbeName = configAccessor.getRequestLocalizedValue("dynamo-db-session-table-name", event);
        const storyMetadata = path.resolve(projectDir, "res", event.request.locale, configAccessor.getRequestLocalizedValue("abc-baked-filename", event));
        const pollyConfig = Object.assign(configAccessor.getRequestLocalizedValue("polly-config", event), {
            s3DomainName: configAccessor.getRequestLocalizedValue("s3-domain-name", event),
            bucketName: configAccessor.getRequestLocalizedValue("s3-bucket-name", event),
            workingDir: path.resolve("/", "tmp"),
            FFMPEGLocation: path.resolve("/", "tmp", SFBRequestHandlerFactory.FFMPEG_BIN_NAME)
        });
        this.setupFFMPEGbin(path.resolve(projectDir, SFBRequestHandlerFactory.FFMPEG_BIN_NAME), pollyConfig.workingDir);
        const defaultPollyConfig = configAccessor.getRequestLocalizedValue("default-narrator", event);
        if (!customRequestHandler) {
            customRequestHandler = AlexaSFBRequestHandler_1.SFBRequestHandler;
        }
        const sfbHandler = new customRequestHandler({
            locale: locale,
            story: require(storyMetadata),
            pollyVoiceConfig: pollyConfig,
            defaultVoiceConfig: defaultPollyConfig,
            attributeTableName: dynamoDBTalbeName
        }, extensions, configAccessor, projectDir);
        return sfbHandler;
    }
    static setupFFMPEGbin(ffmpegSrcPath, workingDir) {
        if (!fs.existsSync(ffmpegSrcPath)) {
            console.log(`[WARN] Packaged FFMPEG not found in path: ${ffmpegSrcPath}`);
            return;
        }
        if (!fs.existsSync(workingDir)) {
            console.log(`[INFO] Initializing audio working directory: ${workingDir}`);
            fs.mkdirSync(workingDir);
        }
        if (!fs.existsSync(workingDir)) {
            console.log(`[WARN] Audio working directory initialization failed!`);
            return;
        }
        const destPath = path.join(workingDir, SFBRequestHandlerFactory.FFMPEG_BIN_NAME);
        if (fs.existsSync(destPath)) {
            console.log(`[INFO] FFMPEG already exists in the audio working directory: ${workingDir}`);
            return;
        }
        try {
            console.log(`[INFO] Copying FFMPEG binary to audio working directory...`);
            fs.copyFileSync(ffmpegSrcPath, destPath);
            fs.chmodSync(destPath, '0700');
        }
        catch (err) {
            throw err;
        }
        return;
    }
}
exports.SFBRequestHandlerFactory = SFBRequestHandlerFactory;
SFBRequestHandlerFactory.FFMPEG_BIN_NAME = "ffmpeg";
