// @alexa-games/sfb-f/dist/test/unitTests/parsers/tokenReplacer.perf-spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tokenReplacer_1 = require("../../../transformers/tokenReplacer");
const segmenter_1 = require("../../../transformers/segmenter");
const legacy_1 = require("../../../transformers/legacy");
// Segmenter is stateless and so can be used globally.
const data = {
    one: 1,
    two: 2,
    threestr: 'its 3rd',
    complex: {
        four: 4,
        five: 5,
        level2: {
            six: 6
        },
        loop: '{complex.loop}'
    },
    level1: {
        level2: {
            level3: {
                data: 'Test value goes here',
                data1: 'More test data.'
            }
        }
    },
    loop: '{loop}',
    location1: 'Ballard',
    location2: 'Seattle',
    index: 2,
    pick1: 2,
    pick2: 1
};
describe("TokenReplacer Performance", function () {
    // *say this will cause a infinite {loop}
    // *say you are at {location{index}}
    const tokenReplacer = new tokenReplacer_1.TokenReplacer(segmenter_1.SegmenterBuilder.getVariableSegmenter(), data);
    it("performance test new parser.", function () {
        let input = '';
        for (let i = 0; i < 1; i++) {
            input += 'More {two} text to {complex.level2.six} data to {two} expand. {\'level1.\'level2.level3.data}, with {missing}. Move to {location{pick{index}}}. ';
        }
        const count = 500000;
        const newStartTime = Date.now();
        for (let i = 0; i < count; i++) {
            tokenReplacer.replaceTokens(input, false);
        }
        const newEndTime = Date.now();
        console.log(`New Method Time: ${newEndTime - newStartTime}`);
        const oldStartTime = Date.now();
        for (let i = 0; i < count; i++) {
            legacy_1.substituteSlotValues(input, data, false);
        }
        const oldEndTime = Date.now();
        console.log(`Original Method Time: ${oldEndTime - oldStartTime}`);
    });
    it("common performance test new parser.", function () {
        let input = '';
        for (let i = 0; i < 1; i++) {
            input += 'This {one} and that {threestr}. ';
        }
        const count = 500000;
        const newStartTime = Date.now();
        for (let i = 0; i < count; i++) {
            tokenReplacer.replaceTokens(input, false);
        }
        const newEndTime = Date.now();
        console.log(`New Method Time: ${newEndTime - newStartTime}`);
        const oldStartTime = Date.now();
        for (let i = 0; i < count; i++) {
            legacy_1.substituteSlotValues(input, data, false);
        }
        const oldEndTime = Date.now();
        console.log(`Original Method Time: ${oldEndTime - oldStartTime}`);
    });
    it("long string performance test new parser.", function () {
        let input = '';
        for (let i = 0; i < 200; i++) {
            input += 'More {two} text to {complex.level2.six} data to {two} expand. {\'level1.\'level2.level3.data}, with {missing}. Move to {location{pick{index}}}. ';
        }
        const count = 5000;
        const newStartTime = Date.now();
        for (let i = 0; i < count; i++) {
            tokenReplacer.replaceTokens(input, false);
        }
        const newEndTime = Date.now();
        console.log(`New Method Time: ${newEndTime - newStartTime}`);
        const oldStartTime = Date.now();
        for (let i = 0; i < count; i++) {
            legacy_1.substituteSlotValues(input, data, false);
        }
        const oldEndTime = Date.now();
        console.log(`Original Method Time: ${oldEndTime - oldStartTime}`);
    });
});
