// @alexa-games/sfb-f/dist/importPlugins/storyFeatureEnumerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const storyBlockFinder_1 = require("./storyBlockFinder");
const storyBlock_1 = require("./storyBlock");
class SceneEnumerator {
    constructor(lines, sceneId) {
        this.finder = new storyBlockFinder_1.StoryBlockFinder({
            blockEndType: storyBlockFinder_1.EndType.NextBlock,
            lines: lines,
            blockStartMatch: storyBlockFinder_1.BlockStartExpressions.SceneMatch,
            blockName: sceneId
        });
    }
    /**
     * Returns the next scene. Returns undefined when no more scenes are available.
     */
    getNext() {
        const result = this.finder.getNextBlock();
        // TODO: Can we use generics with finder to return SceneStoryBlock rather than
        // transforming here.
        if (result) {
            return new storyBlock_1.SceneStoryBlock(result.lines, result.blockName, result.range);
        }
        return undefined;
    }
}
exports.SceneEnumerator = SceneEnumerator;
class ScenePropertyEnumerator {
    constructor(lines, range, type) {
        this.finder = new storyBlockFinder_1.StoryBlockFinder({
            blockEndType: storyBlockFinder_1.EndType.NextBlock,
            lines: lines,
            range,
            blockName: type,
            blockStartMatch: storyBlockFinder_1.BlockStartExpressions.ScenePropMatch
        });
    }
    /**
     * Returns the next scene property. Returns undefined when no more are available.
     */
    getNext() {
        const result = this.finder.getNextBlock();
        // TODO: Can we use generics with finder to return SceneStoryBlock rather than
        // transforming here.
        if (result) {
            return new storyBlock_1.ScenePropertyStoryBlock(result.lines, result.blockName, result.range);
        }
        return undefined;
    }
}
exports.ScenePropertyEnumerator = ScenePropertyEnumerator;
