// @alexa-games/sfb-skill/dist/test/aceExtension/testUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sfb_f_1 = require("@alexa-games/sfb-f");
const sfb_f_2 = require("@alexa-games/sfb-f");
exports.testUserInput = {
    intent: "something",
    slots: [
        {
            name: "slotName",
            value: "value"
        }
    ],
    value: "raw value"
};
function getDefaultBaseDriver() {
    const driver = new sfb_f_1.ACEDriver({
        storyTitle: 'test story',
        storyID: "test-story",
        pluginName: 'default',
        scenes: [{
                id: "start",
                contents: [
                    {
                        narration: "test narration"
                    }
                ]
            }]
    });
    return driver;
}
exports.getDefaultBaseDriver = getDefaultBaseDriver;
function getMockInstructionParameter(instruction, params) {
    return { instructionName: instruction,
        instructionParameters: params,
        storyState: {},
        playStage: new sfb_f_1.PlayStage(new sfb_f_2.AudioItemUtil()),
        storyAccessor: new sfb_f_1.StoryAccessor({ pluginName: "",
            storyTitle: "",
            storyID: "",
            scenes: []
        })
    };
}
exports.getMockInstructionParameter = getMockInstructionParameter;
